File: lib/admin-path.ts
/**
 * Public admin base path helpers (WordPress-style hidden admin URL).
 * Internal App Router paths remain under `/admin`; the public URL defaults to `/manage`
 * (and may be a further custom slug when "hide admin path" is enabled).
 */

/** App Router / filesystem base — never change without renaming `app/(admin)/admin`. */
export const INTERNAL_ADMIN_BASE = "/admin";

/** Default public URL for the admin app (what users see in the address bar). */
export const DEFAULT_ADMIN_BASE = "/manage";

export const ADMIN_PATH_COOKIE = "bqi_admin_base";

/** Top-level route segments that must not be used as the admin slug. */
export const RESERVED_ADMIN_PATH_SLUGS = new Set([
  "admin",
  "manage",
  "api",
  "login",
  "sign-up",
  "signup",
  "forgot-password",
  "reset-password",
  "auth",
  "logout",
  "about",
  "contact-us",
  "contact",
  "services",
  "blog",
  "careers",
  "apply",
  "dashboard",
  "employee",
  "employees",
  "offline",
  "status",
  "sitemap",
  "robots",
  "manifest",
  "favicon",
  "icons",
  "images",
  "public",
  "private",
  "temp",
  "draft",
  "teams",
  "sliders",
  "_next",
  "next",
  "static",
  "assets",
  "sw",
  "workbox",
  "monitoring",
  "health",
  "internal",
  "proxy",
  "cdn",
  "www",
]);

const SLUG_PATTERN = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

export type AdminPathConfig = {
  admin_path_hidden: boolean;
  admin_path_slug: string | null;
};

export function normalizeAdminPathSlug(raw: unknown): string | null {
  if (typeof raw !== "string") return null;
  const slug = raw.trim().toLowerCase().replace(/^\/+|\/+$/g, "");
  return slug || null;
}

export function isValidAdminPathSlug(slug: string | null | undefined): boolean {
  if (!slug) return false;
  if (slug.length < 3 || slug.length > 64) return false;
  if (!SLUG_PATTERN.test(slug)) return false;
  if (RESERVED_ADMIN_PATH_SLUGS.has(slug)) return false;
  return true;
}

export function validateAdminPathSlug(raw: unknown): {
  ok: boolean;
  slug: string | null;
  error?: string;
} {
  const slug = normalizeAdminPathSlug(raw);
  if (!slug) {
    return { ok: false, slug: null, error: "Enter a custom URL slug." };
  }
  if (slug.length < 3 || slug.length > 64) {
    return {
      ok: false,
      slug,
      error: "Slug must be 3–64 characters.",
    };
  }
  if (!SLUG_PATTERN.test(slug)) {
    return {
      ok: false,
      slug,
      error: "Use lowercase letters, numbers, and hyphens only.",
    };
  }
  if (RESERVED_ADMIN_PATH_SLUGS.has(slug)) {
    return {
      ok: false,
      slug,
      error: `"${slug}" is reserved. Choose a different slug.`,
    };
  }
  return { ok: true, slug };
}

export function getPublicAdminBasePath(
  config: Partial<AdminPathConfig> | null | undefined
): string {
  const hidden = Boolean(config?.admin_path_hidden);
  const slug = normalizeAdminPathSlug(config?.admin_path_slug);
  if (hidden && slug && isValidAdminPathSlug(slug)) {
    return `/${slug}`;
  }
  return DEFAULT_ADMIN_BASE;
}

function stripKnownAdminPrefix(path: string): string {
  let rest = path;
  for (const prefix of [INTERNAL_ADMIN_BASE, DEFAULT_ADMIN_BASE]) {
    if (rest === prefix || rest.startsWith(`${prefix}/`)) {
      rest = rest.slice(prefix.length);
      break;
    }
  }
  return rest;
}

/** Build a public admin URL from an internal `/admin/...` or `/manage/...` path. */
export function adminHref(
  path: string = "",
  publicBase: string = DEFAULT_ADMIN_BASE
): string {
  const base =
    (publicBase || DEFAULT_ADMIN_BASE).replace(/\/+$/, "") || DEFAULT_ADMIN_BASE;
  let rest = (path || "").trim();
  if (!rest || rest === "/") return base;
  rest = stripKnownAdminPrefix(rest);
  if (!rest.startsWith("/")) rest = `/${rest}`;
  return `${base}${rest === "/" ? "" : rest}`;
}

/** Map a browser pathname back to the internal `/admin/...` path. */
export function toInternalAdminPath(
  pathname: string,
  publicBase: string = DEFAULT_ADMIN_BASE
): string {
  if (!pathname) return pathname;
  const base =
    (publicBase || DEFAULT_ADMIN_BASE).replace(/\/+$/, "") || DEFAULT_ADMIN_BASE;
  if (pathname === INTERNAL_ADMIN_BASE || pathname.startsWith(`${INTERNAL_ADMIN_BASE}/`)) {
    return pathname;
  }
  if (base !== INTERNAL_ADMIN_BASE) {
    if (pathname === base) return INTERNAL_ADMIN_BASE;
    if (pathname.startsWith(`${base}/`)) {
      return `${INTERNAL_ADMIN_BASE}${pathname.slice(base.length)}`;
    }
  }
  if (pathname === DEFAULT_ADMIN_BASE) return INTERNAL_ADMIN_BASE;
  if (pathname.startsWith(`${DEFAULT_ADMIN_BASE}/`)) {
    return `${INTERNAL_ADMIN_BASE}${pathname.slice(DEFAULT_ADMIN_BASE.length)}`;
  }
  return pathname;
}

export function isPublicAdminPath(
  pathname: string,
  publicBase: string = DEFAULT_ADMIN_BASE
): boolean {
  const base =
    (publicBase || DEFAULT_ADMIN_BASE).replace(/\/+$/, "") || DEFAULT_ADMIN_BASE;
  return pathname === base || pathname.startsWith(`${base}/`);
}

export function isInternalAdminPath(pathname: string): boolean {
  return (
    pathname === INTERNAL_ADMIN_BASE ||
    pathname.startsWith(`${INTERNAL_ADMIN_BASE}/`)
  );
}

export function readAdminBasePathCookie(): string | null {
  if (typeof document === "undefined") return null;
  const match = document.cookie
    .split("; ")
    .find((row) => row.startsWith(`${ADMIN_PATH_COOKIE}=`));
  if (!match) return null;
  try {
    const value = decodeURIComponent(match.split("=").slice(1).join("="));
    if (!value.startsWith("/")) return null;
    if (value === DEFAULT_ADMIN_BASE) return DEFAULT_ADMIN_BASE;
    if (value === INTERNAL_ADMIN_BASE) return DEFAULT_ADMIN_BASE;
    const slug = value.slice(1);
    if (!isValidAdminPathSlug(slug)) return null;
    return `/${slug}`;
  } catch {
    return null;
  }
}

export function writeAdminBasePathCookie(publicBase: string) {
  if (typeof document === "undefined") return;
  const base =
    (publicBase || DEFAULT_ADMIN_BASE).replace(/\/+$/, "") || DEFAULT_ADMIN_BASE;
  const maxAge = 60 * 60 * 24 * 365;
  document.cookie = `${ADMIN_PATH_COOKIE}=${encodeURIComponent(base)}; path=/; max-age=${maxAge}; samesite=lax`;
}

export function clearAdminBasePathCookie() {
  if (typeof document === "undefined") return;
  document.cookie = `${ADMIN_PATH_COOKIE}=; path=/; max-age=0; samesite=lax`;
}

File: Backend/app/lib/admin_path.py
"""Admin public path slug helpers (hide /admin behind a custom URL)."""

from __future__ import annotations

import re
from typing import Any, Dict, Optional, Tuple

RESERVED_ADMIN_PATH_SLUGS = {
    "admin",
    "api",
    "login",
    "sign-up",
    "signup",
    "forgot-password",
    "reset-password",
    "auth",
    "logout",
    "about",
    "contact-us",
    "contact",
    "services",
    "blog",
    "careers",
    "apply",
    "dashboard",
    "employee",
    "employees",
    "offline",
    "status",
    "sitemap",
    "robots",
    "manifest",
    "favicon",
    "icons",
    "images",
    "public",
    "private",
    "temp",
    "draft",
    "teams",
    "sliders",
    "_next",
    "next",
    "static",
    "assets",
    "sw",
    "workbox",
    "monitoring",
    "health",
    "internal",
    "proxy",
    "cdn",
    "www",
}

_SLUG_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


def normalize_admin_path_slug(raw: Any) -> Optional[str]:
    if not isinstance(raw, str):
        return None
    slug = raw.strip().lower().strip("/")
    return slug or None


def is_valid_admin_path_slug(slug: Optional[str]) -> bool:
    if not slug:
        return False
    if len(slug) < 3 or len(slug) > 64:
        return False
    if not _SLUG_RE.match(slug):
        return False
    if slug in RESERVED_ADMIN_PATH_SLUGS:
        return False
    return True


def validate_admin_path_slug(raw: Any) -> Tuple[bool, Optional[str], Optional[str]]:
    slug = normalize_admin_path_slug(raw)
    if not slug:
        return False, None, "Enter a custom URL slug."
    if len(slug) < 3 or len(slug) > 64:
        return False, slug, "Slug must be 3–64 characters."
    if not _SLUG_RE.match(slug):
        return False, slug, "Use lowercase letters, numbers, and hyphens only."
    if slug in RESERVED_ADMIN_PATH_SLUGS:
        return False, slug, f'"{slug}" is reserved. Choose a different slug.'
    return True, slug, None


def get_public_admin_base_path(doc: Optional[Dict[str, Any]]) -> str:
    doc = doc or {}
    hidden = bool(doc.get("admin_path_hidden"))
    slug = normalize_admin_path_slug(doc.get("admin_path_slug"))
    if hidden and is_valid_admin_path_slug(slug):
        return f"/{slug}"
    return "/admin"


async def load_admin_path_config(db) -> Dict[str, Any]:
    """Return path config for middleware (never expose via public unauthenticated APIs without a gate)."""
    if db is None:
        return {
            "admin_path_hidden": False,
            "admin_path_slug": None,
            "public_base": "/admin",
        }
    doc = await db.settings.find_one(
        {"type": "admin"},
        {"admin_path_hidden": 1, "admin_path_slug": 1},
    )
    hidden = bool((doc or {}).get("admin_path_hidden"))
    slug = normalize_admin_path_slug((doc or {}).get("admin_path_slug"))
    if not is_valid_admin_path_slug(slug):
        slug = None
        hidden = False
    return {
        "admin_path_hidden": hidden and bool(slug),
        "admin_path_slug": slug if hidden else slug,
        "public_base": get_public_admin_base_path(
            {"admin_path_hidden": hidden, "admin_path_slug": slug}
        ),
    }


def sanitize_admin_path_fields(payload: Dict[str, Any]) -> Dict[str, Any]:
    """Validate and normalize admin path fields on settings update. Raises ValueError."""
    out = dict(payload)
    touching = "admin_path_hidden" in out or "admin_path_slug" in out
    if not touching:
        return out

    hidden = bool(out.get("admin_path_hidden", False))
    raw_slug = out.get("admin_path_slug")
    slug = normalize_admin_path_slug(raw_slug)

    if hidden:
        ok, slug, err = validate_admin_path_slug(slug)
        if not ok:
            raise ValueError(err or "Invalid admin path slug.")
        out["admin_path_hidden"] = True
        out["admin_path_slug"] = slug
    else:
        out["admin_path_hidden"] = False
        if slug is None or slug == "":
            out["admin_path_slug"] = None
        else:
            ok, slug, err = validate_admin_path_slug(slug)
            if not ok:
                raise ValueError(err or "Invalid admin path slug.")
            out["admin_path_slug"] = slug

    return out
